urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/CustomUtil.java

```diff
@@ -0,0 +1,536 @@
+     * quoted_string := '"' <any character sequence without quotes> '"' | "'" <any character sequence without quotes> "'"
+     *
+     * Example: 
+     * username="keel", realm="Password Required", nonce="123456", uri="/secret", algorithm=MD5, response="c21340f3fdd7194ab8e397ff1a9a750f"
+     *
+     * The returned map will contain the parsed key-tuple-pairs. The key comparator is not case-sensitive.
+     *
+     **/
+    public static Map<String,String> parseDigestAuthorizationChallenge( String line ) {
+
+	Map<String,String> map = new TreeMap<String,String>( CaseInsensitiveComparator.sharedInstance );
+
+	// This regex works but includes the quotes in the final result.
+	// The line will be splitted into the single '<key>=<value>' elements by this regex.
+	Pattern regex = Pattern.compile( "[^\\s]+=([^\\s\"']+|\"[^\"]*\"|'[^']*')" );
+
+
+	Matcher regexMatcher = regex.matcher( line );
+	while( regexMatcher.find() ) {
+
+	    String token = regexMatcher.group();
+
+	    // Remove comma?
+	    if( token.endsWith(",") )
+		token = token.substring(0,token.length()-1);
+
+	    // Split at '='
+	    int index = token.indexOf("=");
+	    if( index == -1 ) {
+
+		// A single '<key>' (without '=<value>')
+		map.put( token, null );
+
+	    } else if( index+1 >= token.length() ) {
+
+		// A '<key>' with '=' but without '<value>'
+		map.put( token.substring(0,index).trim(), "" );
+
+	    } else {
+
+		// A full '<key>=<value>' pair.
+
+		String key   = token.substring(0,index).trim();
+		String value = token.substring(index+1,token.length()).trim();
+		
+		// Remove quotes?	    
+		/*if( value.length() >= 2 
+		    && ( ( value.startsWith("\"") && value.endsWith("\"") )
+			 ||
+			 ( value.startsWith("'") && value.endsWith("'") ) 
+			 ) ) {
+
+		    value = value.substring( 1, value.length()-1 );
+		}
+		*/
+		value = stripQuotes( value );
+
+		map.put( key, value );
+	    
+	    }
+
+
+	} 
+	
+	return map;
+    }
+
+    public static String stripQuotes( String str ) {
+	if( str == null )
+	    return null;
+	
+	if( str.length() >= 2 
+	    && ( ( str.startsWith("\"") && str.endsWith("\"") )
+		 ||
+		 ( str.startsWith("'") && str.endsWith("'") ) 
+		 ) ) {
+	    
+	    return str.substring( 1, str.length()-1 );
+	} else {
+	    return str;
+	}
+    }
+
+    /*public static String[] splitAtWhitespace( String str ) {
+	if( str == null )
+	    return null;
+	return str.split( "\\s+" );
+	}*/
+
+    public static int[] string2int( String[] splits ) 
+	throws NullPointerException,
+	       NumberFormatException {
+
+	if( splits == null )
+	    return null;
+	int[] arr = new int[ splits.length ];
+	for( int i = 0; i < splits.length; i++ ) {
+
+	    if( splits[i] == null )
+		throw new NullPointerException( "Token at index " + i + " is null and cannot be converted to int." );
+	    
+	    try {
+		arr[i] = Integer.parseInt( splits[i].trim() );
+	    } catch( NumberFormatException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)